synapse/storage/databases/main/deviceinbox.py

```diff
@@ -136,63 +137,241 @@ def process_replication_rows(self, stream_name, instance_name, token, rows):
-            # user/device pair up to current_stream_id.
-            if len(messages) < limit:
-                stream_pos = current_stream_id
+    async def _get_device_messages(
+        self,
+        user_ids: Collection[str],
+        from_stream_id: int,
+        to_stream_id: int,
+        device_ids: Optional[Collection[str]] = None,
+        limit: Optional[int] = None,
+    ) -> Tuple[Dict[Tuple[str, str], List[JsonDict]], int]:
+        """
+        Retrieve pending to-device messages for a collection of user devices.
 
-            return messages, stream_pos
+        Only to-device messages with stream ids between the given boundaries
+        (from < X <= to) are returned.
+
+        Note that a stream ID can be shared by multiple copies of the same message with
+        different recipient devices. Stream IDs are only unique in the context of a single
+        user ID / device ID pair Thus, applying a limit (of messages to return) when working
+        with a sliding window of stream IDs is only possible when querying messages of a
+        single user device.
+
+        Finally, note that device IDs are not unique across users.
+
+        Args:
+            user_ids: The user IDs to filter device messages by.
+            from_stream_id: The lower boundary of stream id to filter with (exclusive).
+            to_stream_id: The upper boundary of stream id to filter with (inclusive).
+            device_ids: If provided, only messages destined for these device IDs will be returned.
+                If not provided, all device IDs for the given user IDs will be used.
+            limit: The maximum number of to-device messages to return. Can only be used when
+                passing a single user ID / device ID tuple.
+
+        Returns:
+            A tuple containing:
+                * A dict of (user_id, device_id) -> list of to-device messages
+                * The last-processed stream ID. If this is less than `to_stream_id`, then
+                    there may be more messages to retrieve. If `limit` is not set, then this
+                    is always equal to 'to_stream_id'.
+        """
+        # A limit can only be applied when querying for a single user ID / device ID tuple.
+        if limit:
+            if not device_ids:
+                raise AssertionError(
+                    "Programming error: _get_new_device_messages was passed 'limit' "
+                    "but not device_ids. This could lead to querying multiple user ID "
+                    "/ device ID pairs, which is not compatible with 'limit'"
+                )
+
+            if len(user_ids) > 1 or len(device_ids) > 1:
+                raise AssertionError(
+                    "Programming error: _get_new_device_messages was passed 'limit' "
+                    "with >1 user id/device id pair"
+                )
+
+        user_ids_to_query: Set[str] = set()
+        device_ids_to_query: Set[str] = set()
+
+        if device_ids is not None:
```

Comment:
seems like a non-empty `device_ids` list should only be permitted when there is exactly one `user_id` (since device_ids are only unique to a given user, and you might otherwise end up querying a given device_id for the wrong user)

Reply:
https://github.com/matrix-org/synapse/pull/11215/commits/089e04168f64be088b89cb696dab76eb8b9f68ad tightens up the interface.